src/mongo_odm/test/query_builder.cpp

```diff
@@ -0,0 +1,137 @@
+// Copyright 2016 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "catch.hpp"
+
+#include <iostream>
+
+#include <bsoncxx/builder/stream/document.hpp>
+#include <mongocxx/client.hpp>
+#include <mongocxx/instance.hpp>
+#include <mongocxx/pipeline.hpp>
+#include <mongocxx/stdx.hpp>
+
+#include <mongo_odm/odm_collection.hpp>
+#include <mongo_odm/query_builder.hpp>
+
+using namespace mongocxx;
+using namespace mongo_odm;
+
+class Bar {
+   public:
+    Bar *v;
+    int64_t w;
+    int x1;
+    int x2;
+    bool y;
+    std::string z;
+    ODM_MAKE_KEYS(Bar, NVP(v), NVP(w), NVP(x1), NVP(x2), NVP(y), NVP(z))
```

Comment:
The `Bar *v` is something I forgot to remove earlier from the tests, I'll delete it. 

The C Preprocessor doesn't let you specify individual arguments in an argument list, so we wouldn't be able to do "`NVP(x)` for each `x`". 

However, the `ODM_MAKE_KEYS` macro creates a tuple for the member pointers, it might be possible to write a `serialize` method that iterates over that tuple when called, and use `this->*nvp` to get the value, for each `nvp` in `fields`.